dataset3.1.md
%dataset3(1) user manual | version 3.0.0-alpha 311aa47
% R. S. Doiel and Tom Morrell
% 2025-05-28

# NAME

dataset3 

# SYNOPSIS

dataset3 [GLOBAL_OPTIONS] VERB [OPTIONS] COLLECTION_NAME [PRAMETER ...]

# DESCRIPTION

dataset3 command line interface supports creating JSON object
collections and managing the JSON object documents in a collection.

When creating new documents in the collection or updating documents
in the collection the JSON source can be read from the command line,
a file or from standard input.

# SUPPORTED VERBS

help
: will give documentation of help on a verb, e.g. "help create"

init C_NAME
: Initialize a new dataset collection named with C_NAME.

create [OPTION] C_NAME KEY
: creates a new JSON document in the collection

read [OPTION] C_NAME KEY
: retrieves the "current" version of a JSON document from 
  the collection writing it standard out

update [OPTION] C_NAME KEY
: updates a JSON document in the collection

delete C_NAME KEY
: removes all versions of a JSON document from the collection

keys [OPTION] C_NAME
: returns a list of keys in the collection

codemeta C_NAME [PATH_TO_NEW_CODEMETA_JSON]
: displays an existing codemetada.json file for the collections or
if an optional path to a new codemeta.json it copies the file and updates the 
collections metadata.

dump C_NAME
: This will write out all dataset collection records in a JSONL document.
JSONL shows on JSON object per line, see https://jsonlines.org for details.
The object rendered will have two attributes, "key" and "object". The
key corresponds to the dataset collection key and the object is the JSON
value retrieved from the collection.

load [OPTION] C_NAME
: This will read JSON objects one per line from standard input. This
format is often called JSONL, see https://jsonlines.org. The object
has two attributes, key and object. 

A word about "keys". dataset3 uses the concept of key/values for
storing JSON documents where the key is a unique identifier and the
value is the object to be stored.  Keys are composed as lower case 
alpha numeric characters but may include period, dash and underscore.
While keys maybe provided in upper and lower case they are always
converted to lowercase internally.

There are three "GLOBAL_OPTIONS" in v3 of dataset3.T hey are 
`-version`, `-help`
and `-license`. All other options come
after the verb and apply to the specific action the verb
implements.

# STORAGE TYPE

There are currently three support storage options for JSON documents in a dataset collection.

- SQLite3 database (default),

The following storage engines were removed in v3 -- pairtree, MySQL and PostgreSQL. If you need
to migrate data from a v2 dataset instance use the dump verb. Then you can import the data
into the v3 dataset collection using the load verb.

# OPTIONS

-help
: display help

-license
: display license

-version
: display version

# EXAMPLES

~~~
   dataset3 help init

   dataset3 init my_objects.ds 

   dataset3 help create

   dataset3 create my_objects.ds "123" '{"one": 1}'

   dataset3 create my_objects.ds "234" mydata.json 
   
   cat <<EOT | dataset3 create my_objects.ds "345"
   {
	   "four": 4,
	   "five": "six"
   }
   EOT

   dataset3 update my_objects.ds "123" '{"one": 1, "two": 2}'

   dataset3 delete my_objects.ds "345"

   dataset3 keys my_objects.ds
~~~

This is an example of initializing a JSON documentation
collection.

~~~
dataset3 init '${C_NAME}'
~~~

In this case '${C_NAME}' is the name of your JSON document
read from the environment varaible C_NAME.

dataset3 3.0.0-alpha




CHANGES.md
Releasse 3.0.0:

This release contains breaking changes with v2 and earlier. It comes with two programs, `dataset3` and `dataset3d`. The verbs used by `dataset3` are limited to the following

- init
- help
- create
- read
- update
- delete
- keys
- query

The end points for `dataset3d` are similar to v2 but without endpoints related to frames and attachments.

V3 does not support pairtree storage. The default storage uses SQLite3.

Release 2.0.0-b06:

This release is a rewrite of version 1 focusing on removing features, clearifying useful concepts and abstracting the storage engines cleanly. The latter was done to allow the web implementation of dataset to achieve an appropriate performance and be able to scale to a larger number of collections and size of collections.


The dataset collection's structure has changed.

- a dataset is a directory containing a collection.json and codemeta.json file
- the collection.json no longer contains general metadata or maps to the keys and pairtree, it focuses on operational settings (e.g. storage type and access information)
    - when running dataset as a web service or on a shared user machine you can setup the database connection through the environment. I.e. set DATASET_DSN_URI value (DSN URI is formed with a protocol named for the SQL driver, a "://" and the DSN for that driver, e.g. "mysql://DB_USER:DB_PASSWD@/DB_NAME")
- a codemeta.json file is now used for holding general collection level metadata. [codemeta](https://codemeta.github.io/) has been adopted by the data science community for describing data and software
- additional JSON configuration files may be used to manage the collection dependent on storage engine
- Experimental features in dataset v2
    - Using SQL JSON columns to store our JSON documents in a relational database
        - SQLite 3
        - MySQL 8
        - Posgres 14.5 (very experimental, still learning Postgres)
    - Object level versioning 
    - Attachment versioning
    - dataset package to support Async, multi-process access via JSON stored in SQL databases


Golang package changes:

- Minimum Go version is now 1.19
- The dataset v2 package has been substantially reorganized and simplified, most things have changed
    - Collection.Init() now takes two parameters, collection name, an an optional DSN URI, if a DSN URI is provided it'll define the storage engine, e.g. a dataset.SQLSTORE)
    - Collection.DocPath() removed, doesn't make sense anymore since JSON may be stored in a SQL table
    - Collection.Read() only takes two parameters, not three
    - Collection.Keys() returns a list of keys and an error value
    - Collection.KeyExists() was renamed Collection.HasKey() to be more idiomatic in Go
    - Collection.FrameExists() was renamed Collection.HasFrame() to be more idiomatic in Go
    - Collection.Length() returns an int64 rather than an int
    - Collection.MetadataJSON() renamed Collection.Metadata() returns the codemeta JSON for the collection
    - Collection.UpdateMetatada() has been added, takes the name of a codemeta.json to replace the existing codemeta content for the collection.


libdataset:

- The C shared library implementation has been dropped for now do to the challenges of easily cross compiling releases

CLI changes:

- options have be restructured so that most come after the verb
- help has been restructured to better support focusing the help text on the task needed
- the command line version is single user, single process and be default assumes pairtree storage. 
- Experimental features
    - JSON stored in SQL database via JSON column types
        - SQLite3
        - MySQL 8
        - Postgres 14.5


Web Service changes:

- the web service is generally RESTful so the end points no longer map directly to the client syntax
- it is recommended to use SQL storage for your dataset collections explosed using the web service (e.q. SQLite3 or MySQL). This is because a pairtree collection doesn't provide docuemnt locking
- relies on SQL JSON columns for storage
- Access to the SQL storage engine is through either the environment or a URI expressing a storage type as protocol and a data source name to making the connection


Release 1.1.0:

Added attachment support for __datasetd__.

Updated the metadata fields to include richer PersonOrOrg data structures
for author, contributor, funder as well as added an annotation map field for custom metadata.

Added "MetadataJSON()" function for Collection to quickly copy out
the metadata values from a collection.

```
    c, err := dataset.Open("MyData.ds")
    ...
    defer c.Close()
    fmt.Printf("%s", c.MetadataJSON())
```

Added "MetadataUpdate()" function to update a collection's metadata.

```
    c, err := dataset.Open("MyData.ds")
    ...
    defer c.Close()
    meta := new(Collection)
    meta.Description = "A test dataset"
    meta.Version = "1.0.0"
    meta.Author = []*PersonOrOrg{
        &Person{
            Type: "Person",
            GivenName: "Jane",
            FamilyName: "Doe",
        },
    }
    err = c.MetadataUpdate(meta)
    ...
```

Depreciated dependency on namaste package and Namaste support in command line tools. Removed "collections.go and collections_test.go" from repository (redundant code). Updated libdataset/libdataset.go to hold functions that were needed for the C-Shared library from collections.go. The Namaste fields in the collection's metadata are now depreciated.

The dataset.Init() now places a lock file in the collection directory and leaves the collection in an "Open" state, it should be explicitly closed after Init is called.

E.g. 

```
   c, err := dataset.Init("MyData.ds")
   ...
   defer c.Close()
```

Removed "set_*" for collection metadata fields from libdataset.go. These should be set using the dataset command line tool only.

The dataset.Analyzer() and dataset.Repair() commands expect the dataset collections to be closed before being called. E.g..

```
    c, err := dataset.Open("MyData.ds")

cmd/dataset3/dataset3.go
// dataset is a command line tool, Go package, shared library and Python package for working with JSON objects as collections on local disc.
//
// @Author R. S. Doiel, <rsdoiel@library.caltech.edu>
//
// Copyright (c) 2025, Caltech
// All rights not granted herein are expressly reserved by Caltech.
//
// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
//
// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
//
// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
package main

import (
	"flag"
	"fmt"
	"os"
	"path"
	"strings"

	// Caltech Library Packages
	"github.com/caltechlibrary/dataset/v3"
)

const (
	helpText = `%{app_name}(1) user manual | version {version} {release_hash}
% R. S. Doiel and Tom Morrell
% {release_date}

# NAME

{app_name} 

# SYNOPSIS

{app_name} [GLOBAL_OPTIONS] VERB [OPTIONS] COLLECTION_NAME [PRAMETER ...]

# DESCRIPTION

{app_name} command line interface supports creating JSON object
collections and managing the JSON object documents in a collection.

When creating new documents in the collection or updating documents
in the collection the JSON source can be read from the command line,
a file or from standard input.

# SUPPORTED VERBS

help
: will give documentation of help on a verb, e.g. "help create"

init C_NAME
: Initialize a new dataset collection named with C_NAME.

create [OPTION] C_NAME KEY
: creates a new JSON document in the collection

read [OPTION] C_NAME KEY
: retrieves the "current" version of a JSON document from 
  the collection writing it standard out

update [OPTION] C_NAME KEY
: updates a JSON document in the collection

delete C_NAME KEY
: removes all versions of a JSON document from the collection

keys [OPTION] C_NAME
: returns a list of keys in the collection

codemeta C_NAME [PATH_TO_NEW_CODEMETA_JSON]
: displays an existing codemetada.json file for the collections or
if an optional path to a new codemeta.json it copies the file and updates the 
collections metadata.

dump C_NAME
: This will write out all dataset collection records in a JSONL document.
JSONL shows on JSON object per line, see https://jsonlines.org for details.
The object rendered will have two attributes, "key" and "object". The
key corresponds to the dataset collection key and the object is the JSON
value retrieved from the collection.

load [OPTION] C_NAME
: This will read JSON objects one per line from standard input. This
format is often called JSONL, see https://jsonlines.org. The object
has two attributes, key and object. 

A word about "keys". {app_name} uses the concept of key/values for
storing JSON documents where the key is a unique identifier and the
value is the object to be stored.  Keys are composed as lower case 
alpha numeric characters but may include period, dash and underscore.
While keys maybe provided in upper and lower case they are always
converted to lowercase internally.

There are three "GLOBAL_OPTIONS" in v3 of {app_name}.T hey are 
` + "`" + `-version` + "`" + `, ` + "`" + `-help` + "`" + `
and ` + "`" + `-license` + "`" + `. All other options come
after the verb and apply to the specific action the verb
implements.

# STORAGE TYPE

There are currently three support storage options for JSON documents in a dataset collection.

- SQLite3 database (default),

The following storage engines were removed in v3 -- pairtree, MySQL and PostgreSQL. If you need
to migrate data from a v2 dataset instance use the dump verb. Then you can import the data
into the v3 dataset collection using the load verb.

# OPTIONS

-help
: display help

-license
: display license

-version
: display version

# EXAMPLES

~~~
   {app_name} help init

   {app_name} init my_objects.ds 

   {app_name} help create

   {app_name} create my_objects.ds "123" '{"one": 1}'

   {app_name} create my_objects.ds "234" mydata.json 
   
   cat <<EOT | {app_name} create my_objects.ds "345"
   {
	   "four": 4,
	   "five": "six"
   }
   EOT

   {app_name} update my_objects.ds "123" '{"one": 1, "two": 2}'

   {app_name} delete my_objects.ds "345"

   {app_name} keys my_objects.ds
~~~

This is an example of initializing a JSON documentation
collection.

~~~
{app_name} init '${C_NAME}'
~~~

In this case '${C_NAME}' is the name of your JSON document
read from the environment varaible C_NAME.

{app_name} {version}

`
)

var (
	// Standard Options
	showHelp    bool
	showLicense bool
	showVersion bool
)

func fmtTxt(src string, appName string, version string) string {
	return strings.ReplaceAll(strings.ReplaceAll(src, "{app_name}", appName), "{version}", version)
}

func main() {
	appName := path.Base(os.Args[0])

	// Standard Options
	flag.BoolVar(&showHelp, "help", false, "display help")
	flag.BoolVar(&showLicense, "license", false, "display license")
	flag.BoolVar(&showVersion, "version", false, "display version")

	// We're ready to process args
	flag.Parse()
	args := flag.Args()

	in := os.Stdin
	out := os.Stdout
	eout := os.Stderr

	if showHelp {
		fmt.Fprintf(out, "%s\n", dataset.FmtHelp(helpText, appName, dataset.Version, dataset.ReleaseDate, dataset.ReleaseHash))
		os.Exit(0)
	}
	if showLicense {
		fmt.Fprintf(out, "%s\n", dataset.LicenseText)
		os.Exit(0)
	}
	if showVersion {
		fmt.Fprintf(out, "%s %s %s\n", appName, dataset.Version, dataset.ReleaseHash)
		os.Exit(0)
	}

	if len(args) == 0 {
		fmt.Fprintf(eout, "%s\n", fmtTxt(helpText, appName, dataset.Version))
		os.Exit(1)
	}

	// Application Logic
	err := dataset.RunCLI(in, out, eout, args)
	if err != nil {
		fmt.Fprintf(eout, "%s\n", err)
		os.Exit(1)
	}
}
